psycopg_c/psycopg_c/types/uuid.pyx

```diff
@@ -0,0 +1,138 @@
+    -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,  // 160-175
+    -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,  // 176-191
+    -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,  // 192-207
+    -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,  // 208-223
+    -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,  // 224-239
+    -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1   // 240-255
+};
+"""
+    const int8_t[256] hex_to_int_map
+
+
+@cython.final
+cdef class UUIDLoader(CLoader):
+    format = PQ_TEXT
+
+    cdef PyObject *_uuid_type
+    cdef object _uuid_new
+    cdef object _obj_setattr
+    cdef PyObject *_safeuuid_unknown
+
+    def __cinit__(self, oid: int, context: AdaptContext | None = None):
+        global uuid
+        # uuid is slow to import, lazy load it
+        if uuid is None:
+            import uuid
+
+        self._uuid_type = <PyObject *>uuid.UUID
+        self._uuid_new = uuid.UUID.__new__
+        self._obj_setattr = object.__setattr__
+        self._safeuuid_unknown = <PyObject *>uuid.SafeUUID.unknown
+
+    cdef object cload(self, const char *data, size_t length):
+        cdef uint64_t high = 0
+        cdef uint64_t low = 0
+        cdef size_t i
+        cdef int ndigits = 0
+        cdef int8_t c
+
+        for i in range(length):
+            c = data[i]
+            if hex_to_int_map[c] == -1:
+                continue
+
+            if ndigits < 16:
+                high = (high << 4) | hex_to_int_map[c]
+            else:
+                low = (low << 4) | hex_to_int_map[c]
+            ndigits += 1
+
+        if ndigits != 32:
+            raise ValueError("Invalid UUID string")
+
+        cdef object int_value = (PyLong_FromUnsignedLongLong(high) << 64) | PyLong_FromUnsignedLongLong(low)
+
+        cdef object u = PyObject_CallFunctionObjArgs(self._uuid_new, self._uuid_type, NULL)
+        PyObject_CallFunctionObjArgs(self._obj_setattr, <PyObject *>u, <PyObject *>"is_safe", self._safeuuid_unknown, NULL)
+        PyObject_CallFunctionObjArgs(self._obj_setattr, <PyObject *>u, <PyObject *>"int", <PyObject *>int_value, NULL)
+        return u
+
+
+@cython.final
+cdef class UUIDBinaryLoader(CLoader):
+    format = PQ_BINARY
+
+    cdef PyObject *_uuid_type
+    cdef object _uuid_new
+    cdef object _obj_setattr
+    cdef PyObject *_safeuuid_unknown
+
+    def __cinit__(self, oid: int, context: AdaptContext | None = None):
+        global uuid
+        # uuid is slow to import, lazy load it
+        if uuid is None:
+            import uuid
+
+        self._uuid_type = <PyObject *>uuid.UUID
+        self._uuid_new = uuid.UUID.__new__
+        self._obj_setattr = object.__setattr__
+        self._safeuuid_unknown = <PyObject *>uuid.SafeUUID.unknown
+
+    cdef object cload(self, const char *data, size_t length):
+        cdef object int_value = int.from_bytes(data[:length], 'big')
```

Comment:
We can load the binary data into an array of two 64 bits uuid, swap them to host order, and then use the same high+low procedure used in the text loader to get a 128 bits integer.

```suggestion
        cdef uint64_t be[2]
        if length != sizeof(be):
            raise ValueError("Invalid UUID data")
        memcpy(&be, data, sizeof(be))

        cdef uint64_t high = endian.be64toh(be[0])
        cdef uint64_t low = endian.be64toh(be[1])
        cdef object int_value = (PyLong_FromUnsignedLongLong(high) << 64) | PyLong_FromUnsignedLongLong(low)
```

Reply:
I tested this but it wasn't any faster than just using `int.from_bytes`, so I opted for a the option that is less code. The two 64 bit numbers approach wasn't faster in the case of `UUIDLoader` either but it seemed less error prone that what I initially proposed, I mentioned it [here](https://github.com/psycopg/psycopg/pull/998#discussion_r1938580621). Let me know if you think we need to implement it anyway.

Reply:
Wasn't the conversation you linked about the text loader? It seems strange that 16 bytes of memcopy are faster than all the Python bureaucracy necessary to call and to execute [that method](https://github.com/python/cpython/blob/80b9e79d84e835ecdb5a15c9ba73e44803ca9d32/Objects/clinic/longobject.c.h#L397).

So I think this is more an error in the measurement and I would rather use the memcpy implementation of this method :slightly_smiling_face: